commons/zenoh-buffers/src/lib.rs

```diff
@@ -137,9 +137,14 @@ pub mod writer {
         fn can_write(&self) -> bool {
             self.remaining() != 0
         }
-        /// Provides a buffer of exactly `len` uninitialized bytes to `f` to allow in-place writing.
-        /// `f` must return the number of bytes it actually wrote.
-        fn with_slot<F>(&mut self, len: usize, f: F) -> Result<NonZeroUsize, DidntWrite>
+        /// Provides a buffer of exactly `len` uninitialized bytes to `write` to allow in-place writing.
+        /// `write` must return the number of bytes it actually wrote.
+        ///
+        /// # Safety
+        ///
+        /// Caller must ensure that `write` return an integer lesser than or equal to the length of
+        /// the slice passed in argument
+        unsafe fn with_slot<F>(&mut self, len: usize, write: F) -> Result<NonZeroUsize, DidntWrite>
```

Comment:
I don't think `with_slot` should be marked as `unsafe`. If `with_slot()` is `unsafe` then also `len()` should be marked as `unsafe`?

Reply:
Which `len()` are you talking about?

Reply:
I could implement 
```rust
fn len(&self) -> usize {
    usize::MAX
}
```
and may buffer implementation is broken in the same way. It's a logical bug, not really an `unsafe` implementation.

Reply:
Marking a function `unsafe` or not obey to a simple rule: does calling this function with dumb parameters may corrupt the program? if yes, it must be marked as unsafe.
Current `with_slot` implementation must be marked unsafe.

For `Buffer::len`, we are not in this case, but we may fall into the case of unsafe traits.
For unsafe trait, the rule is: is it possible to implement this trait in a way that may corrupt the program? if yes, the trait must be marked unsafe.
I have not looked in detail to the `Buffer` trait yet, but maybe it should be marked as unsafe too.

Reply:
Calling `len()` on a slice is not `unsafe` but clearly a wrong implementation `len()` will lead to a buggy behaviour. In any case, the wrong behaviour may be a well-defined and not undefined behaviour. Calling `with_slot()` with wrong parameters will lead to a buggy behaviour but it will be deterministic behaviour, in the same way an `unwrap()` may crash your program. Therefore it doesn't need to be `unsafe` for me.

Reply:
It's not about undeterministic behavior, it's about memory corruption and undefined behavior; both don't necessarily make program crash. 

Here is the current implementation of `<&mut [u8]>::with_slot`:
```rust
unsafe fn with_slot<F>(&mut self, len: usize, write: F) -> Result<NonZeroUsize, DidntWrite>
    where
        F: FnOnce(&mut [u8]) -> usize,
    {
        if len > self.len() {
            return Err(DidntWrite);
        }
        let written = write(&mut self[..len]);
        // SAFETY: `written` < `len` is guaranteed by function contract
        *self = unsafe { mem::take(self).get_unchecked_mut(written..) };
        NonZeroUsize::new(written).ok_or(DidntWrite)
    }
 ```
 The reason why `with_slot` must be unsafe is straightforward: the result of `write` is used with `get_unchecked_mut`, and `get_unchecked_mut` has a precondition that must be respected. The only way to do it without changing the implementation is to add a precondition to `with_slot`.

Reply:
> Calling len() on a slice is not unsafe but clearly a wrong implementation len() will lead to a buggy behaviour.

You need to define "buggy". If buggy means undefined behavior or memory corruption because other parts of the code rely on `Buffer::len` result to do unsafe things, then the trait must be marked unsafe.

Reply:
About unsafe trait, you may look at `ExactSizeIterator` [documentation](https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html):
> Note that this trait is a safe trait and as such does not and cannot guarantee that the returned length is correct. This means that unsafe code must not rely on the correctness of [Iterator::size_hint](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.size_hint). The unstable and unsafe [TrustedLen](https://doc.rust-lang.org/std/iter/trait.TrustedLen.html) trait gives this additional guarantee.

So, if you use `Buffer::len` in any unsafe precondition, you should mark `Buffer` unsafe, especially it is public.

Reply:
A proposal to make `with_slot` safe would be the following:
```rust
   fn with_slot<F>(&mut self, len: usize, write: F) -> Result<NonZeroUsize, DidntWrite>
    where
        F: FnOnce(&mut [u8]) -> &[u8],
    {
        if len > self.len() {
            return Err(DidntWrite);
        }
        let written = write(&mut self[..len]);
        assert_eq!(self.as_ptr(), written.as_ptr());
        let written_length = written.len();
        // SAFETY: Assertion ensures that `self` and `written` are the same slice, 
        // so `written.len()` is a valid length for `self`
        *self = unsafe { mem::take(self).get_unchecked_mut(written_length..) };
        NonZeroUsize::new(written_length).ok_or(DidntWrite)
    }
```
Here, the safety is ensured by an assertion, there is no need of `unsafe` precondition.
But because `write` callback should be inlined, the assertion should be elided. I have to test it.
Now, the only way to trigger memory corruption by calling `with_slot` would be to use `unsafe` code to forge a returned slice with the same base pointer and an overflowing length. However, the fault would be on the user `unsafe` code, not on `with_slot`, so it doesn't question `with_slot` safety.